File: test_py/test_make_request.py
import pytest
from responses import RequestsMock
from responses.matchers import header_matcher, json_params_matcher, query_param_matcher
from responses.registries import OrderedRegistry
from src_py.grndwork_api_client.config import API_URL
from src_py.grndwork_api_client.make_request import make_request, RequestError

TEST_URL = f'{API_URL}/v1/test'


def describe_make_request():
    @pytest.fixture(name='api_mock', autouse=True)
    def fixture_api_mock():
        with RequestsMock(registry=OrderedRegistry) as api_mock:
            yield api_mock

    def it_makes_request_with_token(api_mock):
        api_mock.get(
            url=TEST_URL,
            match=[header_matcher({'Authorization': 'Bearer auth_token'})],
            status=200,
            json={},
        )

        make_request(
            url=TEST_URL,
            token='auth_token',
        )

    def it_makes_request_with_additional_headers(api_mock):
        api_mock.get(
            url=TEST_URL,
            match=[header_matcher({'X-Test': 'test_value'})],
            status=200,
            json={},
        )

        make_request(
            url=TEST_URL,
            headers={
                'X-Test': 'test_value',
            },
        )

    def it_makes_request_with_query(api_mock):
        api_mock.get(
            url=TEST_URL,
            match=[query_param_matcher({'limit': 10})],
            status=200,
            json={},
        )

        make_request(
            url=TEST_URL,
            query={'limit': 10},
        )

    def it_makes_request_with_method(api_mock):
        api_mock.post(
            url=TEST_URL,
            status=201,
            json={},
        )

        make_request(
            url=TEST_URL,
            method='POST',
        )

    def it_makes_request_with_body(api_mock):
        api_mock.post(
            url=TEST_URL,
            match=[json_params_matcher({'test': 'value'})],
            status=201,
            json={},
        )

        make_request(
            url=TEST_URL,
            method='POST',
            body={
                'test': 'value',
            },
        )

    def it_raises_error_when_bad_request(api_mock):
        api_mock.get(
            url=TEST_URL,
            status=400,
            json={},
        )

        with pytest.raises(RequestError, match='Bad Request'):
            make_request(
                url=TEST_URL,
            )

    def it_raises_error_with_response_body(api_mock):
        api_mock.get(
            url=TEST_URL,
            status=400,
            json={'message': 'Invalid'},
        )

        with pytest.raises(RequestError, match='Invalid'):
            make_request(
                url=TEST_URL,
            )

    def it_raises_error_when_bad_response_body(api_mock):
        api_mock.get(
            url=TEST_URL,
            status=200,
            body='Invalid',
        )

        with pytest.raises(RequestError, match='Failed to parse response payload'):
            make_request(
                url=TEST_URL,
            )

    def it_returns_payload_and_response(api_mock):
        api_mock.get(
            url=TEST_URL,
            status=200,
            headers={
                'X-Test': 'test_value',
            },
            json={'test': 'value'},
        )

        payload, resp = make_request(
            url=TEST_URL,
        )

        assert payload == {'test': 'value'}
        assert resp['status_code'] == 200
        assert resp['headers'].get('x-test') == 'test_value'

File: test_js/test_make_request.ts
import * as undici from 'undici';
import {API_URL} from '../src_js/grndwork_api_client/config';
import {makeRequest, RequestError} from '../src_js/grndwork_api_client/make_request';

const TEST_URL = `${API_URL}/v1/test`;

describe('makeRequest', () => {
  const TEST_PATH = new URL(TEST_URL).pathname;

  let globalAgent: undici.Dispatcher;
  let mockAgent: undici.MockAgent;
  let apiMock: undici.MockPool;

  beforeEach(() => {
    globalAgent = undici.getGlobalDispatcher();

    mockAgent = new undici.MockAgent();
    mockAgent.disableNetConnect();
    undici.setGlobalDispatcher(mockAgent);

    apiMock = mockAgent.get(API_URL);
  });

  afterEach(async () => {
    undici.setGlobalDispatcher(globalAgent);
    mockAgent.assertNoPendingInterceptors();
    await mockAgent.close();
  });

  it('makes request with token', async () => {
    apiMock.intercept({
      path: TEST_PATH,
      headers: {Authorization: 'Bearer auth_token'},
    })
    .reply(200, {});

    await makeRequest({
      url: TEST_URL,
      token: 'auth_token',
    });
  });

  it('makes request with additional headers', async () => {
    apiMock.intercept({
      path: TEST_PATH,
      headers: {'X-Test': 'test_value'},
    })
    .reply(200, {});

    await makeRequest({
      url: TEST_URL,
      headers: {
        'X-Test': 'test_value',
      },
    });
  });

  it('makes request with query', async () => {
    apiMock.intercept({
      path: TEST_PATH,
      query: {limit: 10},
    })
    .reply(200, {});

    await makeRequest({
      url: TEST_URL,
      query: {limit: 10},
    });
  });

  it('makes request with method', async () => {
    apiMock.intercept({
      path: TEST_PATH,
      method: 'POST',
    })
    .reply(201, {});

    await makeRequest({
      url: TEST_URL,
      method: 'POST',
    });
  });

  it('makes request with body', async () => {
    apiMock.intercept({
      path: TEST_PATH,
      method: 'POST',
      body: JSON.stringify({test: 'value'}),
    })
    .reply(201, {});

    await makeRequest({
      url: TEST_URL,
      method: 'POST',
      body: {
        test: 'value',
      },
    });
  });

  it('throws error when bad request', async () => {
    apiMock.intercept({
      path: TEST_PATH,
    })
    .reply(400, {});

    await expect(
      () => makeRequest({
        url: TEST_URL,
      }),
    ).rejects.toThrow(new RequestError('Bad Request'));
  });

  it('throws error with response body', async () => {
    apiMock.intercept({
      path: TEST_PATH,
    })
    .reply(400, {message: 'Invalid'}, {
      headers: {
        'content-type': 'application/json',
      },
    });

    await expect(
      () => makeRequest({
        url: TEST_URL,
      }),
    ).rejects.toThrow(new RequestError('Invalid'));
  });

  it('throws error when bad response body', async () => {
    apiMock.intercept({
      path: TEST_PATH,
    })
    .reply(200, 'Invalid');

    await expect(
      () => makeRequest({
        url: TEST_URL,
      }),
    ).rejects.toThrow(new RequestError('Failed to parse response payload'));
  });

  it('returns payload and response', async () => {
    apiMock.intercept({
      path: TEST_PATH,
    })
    .reply(200, {test: 'value'}, {
      headers: {
        'X-Test': 'test_value',
      },
    });

    const [payload, resp] = await makeRequest({
      url: TEST_URL,
    });

    expect(payload).toEqual({test: 'value'});
    expect(resp.status_code).toEqual(200);
    expect(resp.headers['x-test']).toEqual('test_value');
  });
});
